Vula

> With zero configuration, Vula automatically encrypts IP communication between hosts on a local area network in a forward-secret and transitionally post-quantum manner to protect against passive eavesdropping.
>
> With manual key verification and/or automatic key pinning and manual resolution of IP or hostname conflicts, Vula will additionally protect against interception by active adversaries.
> 
> When the local gateway to the internet is also Vula peer, internet-destined traffic will also be encrypted on the LAN.

- Website: https://vula.link/
- source: https://codeberg.org/vula/vula
- installation instructions: https://codeberg.org/vula/vula/src/branch/main/INSTALL.md
- hacking instructions: https://codeberg.org/vula/vula/src/branch/main/HACKING.md

Wanted to update the team regarding where @GetPsyched and I left things today while wrapping up pairing with @ioerror:

Running the vula nixosTest:

```
$ nix build .#hydraJobs.tests.x86_64-linux.nixos.Vula.vula.driverInteractive -L
$ ./result/bin/nixos-test-driver
...
>>> start_all()
...
>>> run_tests()
...
```

Resulted in output from vula regarding an invalid descriptor:

```
...
a # [   54.231369] vula[797]: 2024-05-16-15:00:24: discover dropped invalid descriptor: {'r': '', 'pk': b'Zv9xR32f0FArb6wTLPcfHus/vQcrX3/fv9uRfMfbAhM=', 'c': b'qopkW/9Ih6KSNPrLJqIfcly5MwNqXYSA129Sr7BIbVqvfECJQwXHckXfL1Uq+3Sioh1maOypOZGrp71LRWJNYg==', 'addrs': b'192.168.1.1', 'vk': b'NVx3TrxYuuuytO0Ptc9wXK0bh9j54CAts3KDAkpJTl4=', 'vf': b'1715871624', 'dt': b'86400', 'port': b'5354', 'hostname': b'a.local.', 'e': b'False', 's': b'ZlcV8uUl5ItRanpKNYkI5SHx2xa9eI9OzwclFfObEm8AcrChuKwxLYyiTM92gxAgAm4cgL0GjhD54OpWUueXDg=='} (SchemaError("b'Zv9xR32f0FArb6wTLPcfHus/vQcrX3/fv9uRfMfbAhM=' is not 32 bytes or a 44-char base64 string which decodes to 32 bytes"))
...
```

Even though checking the SchemaError value of `pk` seems valid and is able to be decoded:

```
$ python3
Python 3.11.9 (main, Apr  2 2024, 08:25:04) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> z = 'Zv9xR32f0FArb6wTLPcfHus/vQcrX3/fv9uRfMfbAhM='
>>> len(z)
44
>>> import base64
>>> base64.b64.decode(z)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'base64' has no attribute 'b64'
>>> base64.b64decode(z)
b'f\xffqG}\x9f\xd0P+o\xac\x13,\xf7\x1f\x1e\xeb?\xbd\x07+_\x7f\xdf\xbf\xdb\x91|\xc7\xdb\x02\x13'
>>> len(base64.b64decode(z))
32
```

There were concerns around the python schema package version differences but the version of nixpkgs and vula are the same version `0.7.5`: https://github.com/NixOS/nixpkgs/blob/ee4a6e0f566fe5ec79968c57a9c2c3c25f2cf41d/pkgs/development/python-modules/schema/default.nix#L12 and https://codeberg.org/vula/vula/src/branch/main/Pipfile.lock#L339-L345

We turned to enabling the full unittest suite of vula by passing pytest `--doctest-modules`. This was in hopes that a similar error would with the descriptor would be raised during the tests. The result was 1 test failure across the full test suite:

```
$ nix build .#vula -L
...
vula-unstable-> =================================== FAILURES ===================================
vula-unstable-> _________ TestWireGuardServiceListener.test_add_service_calls_callback _________
vula-unstable-> self = <test.test_discover.TestWireGuardServiceListener object at 0x7ffff28dbbd0>
vula-unstable->     def test_add_service_calls_callback(self):
vula-unstable->         callback = MagicMock()
vula-unstable->         props = {
vula-unstable->             b'addrs': b'192.168.2.1',
vula-unstable->             b'pk': b'EqcQ5gYxzGtzg7B4xi83kLyfuSMp8Kv3cmAJMs12nDM=',
vula-unstable->             b'c': b'T6htsKgwCp5MAXjPiWxtVkccg+K2CePsVa7uyUgxE2ouYxKXg2qNL+'
vula-unstable->             + b'0ut3sSVTYjzFGZSCO6n80SRaR+BIeOCg==',
vula-unstable->             b'hostname': b'myhost',
vula-unstable->             b'port': b'5054',
vula-unstable->             b'vk': b'EqcQ5gYxzGtzg7B4xi83kLyfuSMp8Kv3cmAJMs12nDM=',
vula-unstable->             b'dt': b'34',
vula-unstable->             b'vf': b'42',
vula-unstable->             b'r': b'192.168.2.0/24',
vula-unstable->             b'e': b'1',
vula-unstable->         }
vula-unstable->         zeroconf = MagicMock()
vula-unstable->         zeroconf.get_service_info().properties = props
vula-unstable->         listener = vula.discover.WireGuardServiceListener(callback)
vula-unstable->         listener.add_service(zeroconf, "test_type", "test_name")
vula-unstable-> >       callback.assert_called_once()
vula-unstable-> test/test_discover.py:34:
vula-unstable-> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
vula-unstable-> self = <MagicMock id='140737259795984'>
vula-unstable->     def assert_called_once(self):
vula-unstable->         """assert that the mock was called only once.
vula-unstable->         """
vula-unstable->         if not self.call_count == 1:
vula-unstable->             msg = ("Expected '%s' to have been called once. Called %s times.%s"
vula-unstable->                    % (self._mock_name or 'mock',
vula-unstable->                       self.call_count,
vula-unstable->                       self._calls_repr()))
vula-unstable-> >           raise AssertionError(msg)
vula-unstable-> E           AssertionError: Expected 'mock' to have been called once. Called 0 times.
vula-unstable-> /nix/store/lpi16513bai8kg2bd841745vzk72475x-python3-3.11.9/lib/python3.11/unittest/mock.py:918: AssertionError
vula-unstable-> =========================== short test summary info ============================
vula-unstable-> FAILED test/test_discover.py::TestWireGuardServiceListener::test_add_service_calls_callback - AssertionError: Expected 'mock' to have been called once. Called 0 times.
vula-unstable-> ======================== 1 failed, 105 passed in 1.54s =========================
vula-unstable-> /nix/store/558iw5j1bk7z6wrg8cp96q2rx03jqj1v-stdenv-linux/setup: line 1579: pop_var_context: head of shell_variables not a function context
error: builder for '/nix/store/aaj74i50m18vx9774vr317w81dgd3rks-vula-unstable-.drv' failed with exit code 1;
       last 10 log lines:
       >                       self.call_count,
       >                       self._calls_repr()))
       > >           raise AssertionError(msg)
       > E           AssertionError: Expected 'mock' to have been called once. Called 0 times.
       >
       > /nix/store/lpi16513bai8kg2bd841745vzk72475x-python3-3.11.9/lib/python3.11/unittest/mock.py:918: AssertionError
       > =========================== short test summary info ============================
       > FAILED test/test_discover.py::TestWireGuardServiceListener::test_add_service_calls_callback - AssertionError: Expected 'mock' to have been called once. Called 0 times.
       > ======================== 1 failed, 105 passed in 1.54s =========================
       > /nix/store/558iw5j1bk7z6wrg8cp96q2rx03jqj1v-stdenv-linux/setup: line 1579: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/aaj74i50m18vx9774vr317w81dgd3rks-vula-unstable-.drv'.
```

In `TestWireGuardServiceListener.test_add_service_calls_callback` the following lines in this test are suspicious since there doesn't seem to be an actual callback and causes the assertion to fail:
```
listener = vula.discover.WireGuardServiceListener(callback)
listener.add_service(zeroconf, "test_type", "test_name")
```

@ioerror here is the command that I was running from https://github.com/ngi-nix/ngipkgs/tree/mob/vula-test branch to build the vula package and run the tests:

```
$ nix build .#vula -L 
```

Additional Note: after [installing nix](https://nixos.org/download/) make sure to enable flakes since they are needed to use this ngi repo but are experimental:

```
$ echo "experimental-features = nix-command flakes" > ./nix.conf && install -D ./nix.conf $HOME/.config/nix/nix.conf
```

--- a/pkgs/by-name/highctidh/package.nix
+++ b/pkgs/by-name/highctidh/package.nix
@@ -2,39 +2,33 @@
   lib,
   python3,
   fetchgit,
-}:
-with builtins; let
-  python = python3;
+}: let
+  inherit (lib) licenses maintainers;
+
+  version = "1.0.2024060500";
+  src = fetchgit {
+    url = "https://codeberg.org/vula/highctidh";
+    rev = "v${version}";
+    hash = "sha256-TyD5KzUz89RBxsSZeJYOkIzD29DF0BjizpMnsTpFOHI=";
+  };
 in
-  python.pkgs.buildPythonApplication rec {
+  python3.pkgs.buildPythonPackage {
     pname = "highctidh";
-    version = "1.0.2023121800";
-    format = "pyproject";
+    inherit version src;
+    pyproject = true;
 
-    src = fetchgit {
-      url = "https://codeberg.org/vula/highctidh";
-      rev = "v${version}";
-      hash = "sha256-83zTz5iBF/ApJV2hnsT2DfN/T36f73MrXmhLDJa5Z8I=";
-    };
+    sourceRoot = "${src.name}/src";
 
-    postPatch = ''
-      patchShebangs test.sh
-      mkdir -p build/tmp
-    '';
-
-    propagatedBuildInputs = with python.pkgs; [
+    nativeBuildInputs = with python3.pkgs; [
       setuptools
-      build
     ];
 
-    nativeBuildInputs = propagatedBuildInputs;
+    nativeCheckInputs = with python3.pkgs; [pytestCheckHook];
 
-    doCheck = true;
-
-    meta = with lib; {
-      description = "Fork of high-ctidh as as a portable shared library with Python bindings.";
+    meta = {
+      description = "Fork of high-ctidh as as a portable shared library with Python bindings";
       homepage = "https://codeberg.org/vula/highctidh";
       license = licenses.publicDomain;
-      maintainers = with maintainers; [lorenzleutgeb];
+      maintainers = with maintainers; [lorenzleutgeb mightyiam];
     };
   }

--- a/pkgs/by-name/vula/package.nix
+++ b/pkgs/by-name/vula/package.nix
@@ -3,64 +3,55 @@
   python3,
   fetchgit,
   highctidh,
-  coreutils,
-}:
-with builtins; let
-  python = python3;
+  wrapGAppsHook,
+}: let
+  inherit
+    (lib)
+    licenses
+    maintainers
+    ;
 in
-  python.pkgs.buildPythonApplication rec {
+  python3.pkgs.buildPythonApplication {
     pname = "vula";
-    version = "0.2.2023112801";
-    format = "pyproject";
+    version = "unstable-2024-05-17";
 
     src = fetchgit {
       url = "https://codeberg.org/vula/vula";
-      rev = "v${version}";
-      hash = "sha256-hBB6jKCLwgfPsgINuvGuLgihrr9zhG46V6/G0SXdCSc=";
+      rev = "b82933c2d45496afb91727e7ce3dff61ae262473";
+      hash = "sha256-DVjEg28GFmA3fOgXZ8MQ7rwfZtt6WkK1qHnyTnYbKcY=";
     };
 
+    # without removing `pyproject.toml` we don't end up with an executable.
     postPatch = ''
-      substituteInPlace configs/systemd/* \
-        --replace 'ExecStart=vula' "ExecStart=$out/bin/vula"
-
-      substituteInPlace configs/dbus/* \
-        --replace 'Exec=/bin/false' "Exec=${coreutils}/bin/false"
+      rm pyproject.toml
     '';
 
-    propagatedBuildInputs = with python.pkgs;
-      [
-        setuptools
-        pyaudio
-        pyroute2
-        hkdf
-        pynacl
+    propagatedBuildInputs =
+      (with python3.pkgs; [
         click
+        cryptography
+        hkdf
         packaging
-        pyyaml
-        pystray
-        qrcode
         pillow
         pydbus
-        zeroconf
+        pynacl
+        pyroute2
+        pyyaml
+        qrcode
         schema
-        cryptography
         tkinter
-      ]
+        zeroconf
+      ])
       ++ [highctidh];
 
-    postInstall = ''
-      mkdir -p $out/{lib/systemd/system,/share/dbus-1/system-services}
-      cp configs/systemd/* $out/lib/systemd/system/
-      cp configs/dbus/* $out/share/dbus-1/system-services/
-    '';
+    nativeBuildInputs = [wrapGAppsHook];
+    nativeCheckInputs = with python3.pkgs; [pytestCheckHook];
 
-    doCheck = true;
-
-    meta = with lib; {
+    meta = {
       description = "Automatic local network encryption";
       homepage = "https://vula.link/";
       license = licenses.gpl3Only;
-      maintainers = with maintainers; [lorenzleutgeb];
+      maintainers = with maintainers; [lorenzleutgeb mightyiam stepbrobd];
       mainProgram = "vula";
     };
   }

--- a/projects/Vula/dbus.conf.nix
+++ b/projects/Vula/dbus.conf.nix
@@ -1,0 +1,60 @@
+{
+  userPrefix,
+  operatorsGroup,
+}: ''
+  <?xml version="1.0" encoding="UTF-8"?>
+  <!DOCTYPE busconfig PUBLIC "-//freedesktop//DTD D-BUS Bus Configuration 1.0//EN"
+  "http://www.freedesktop.org/standards/dbus/1.0/busconfig.dtd">
+  <busconfig>
+    <type>system</type>
+
+    <policy user="${userPrefix}-organize">
+      <allow own="local.vula.organize"/>
+      <allow
+         send_destination="local.vula.discover"
+         send_interface="local.vula.discover1.Listen"
+         send_type="method_call"
+      />
+      <allow
+         send_destination="local.vula.publish"
+         send_interface="local.vula.publish1.Listen"
+         send_type="method_call"
+      />
+    </policy>
+
+    <policy user="${userPrefix}-publish">
+      <allow own="local.vula.publish"/>
+    </policy>
+
+    <policy user="${userPrefix}-discover">
+      <allow own="local.vula.discover"/>
+      <allow
+         send_destination="local.vula.organize"
+         send_interface="local.vula.organize1.ProcessDescriptor"
+         send_type="method_call"
+      />
+    </policy>
+
+    <policy group="${operatorsGroup}">
+      <allow send_destination="local.vula.organize" />
+    </policy>
+
+    <policy context="default">
+      <allow
+         send_destination="local.vula.organize"
+         send_interface="org.freedesktop.DBus.Introspectable"
+         send_type="method_call"
+      />
+      <allow
+         send_destination="local.vula.publish"
+         send_interface="org.freedesktop.DBus.Introspectable"
+         send_type="method_call"
+      />
+      <allow
+         send_destination="local.vula.discover"
+         send_interface="org.freedesktop.DBus.Introspectable"
+         send_type="method_call"
+      />
+    </policy>
+  </busconfig>
+''

--- a/projects/Vula/default.nix
+++ b/projects/Vula/default.nix
@@ -1,11 +1,13 @@
-{
-  pkgs,
-  lib,
-  sources,
-} @ args: {
+{pkgs, ...} @ args: {
   packages = {inherit (pkgs) vula;};
-  nixos = {
-    modules.services.vula = ./service.nix;
-    tests.vula = import ./test.nix args;
+  nixos.modules.services.vula = ./service.nix;
+  nixos.tests.test = import ./test.nix args;
+  nixos.examples.simple = {
+    path = ./example-simple.nix;
+    description = ''
+      Simple configuration for Vula. Vula nodes will automatically discover each other on networks that support [multicast DNS](https://en.wikipedia.org/wiki/Multicast_DNS) (mDNS).
+
+      Add users to the group defined in `config.services.vula.adminGroup` to grant them permissions to manage Vula through the `vula` command.
+    '';
   };
 }

--- a/projects/Vula/example-simple.nix
+++ b/projects/Vula/example-simple.nix
@@ -1,0 +1,4 @@
+{
+  services.vula.enable = true;
+  services.vula.openFirewall = true;
+}

--- a/projects/Vula/nss-altfiles.nix
+++ b/projects/Vula/nss-altfiles.nix
@@ -1,0 +1,32 @@
+# not exposed because vula uses specific non-release rev and some build flags
+{
+  stdenv,
+  fetchFromGitHub,
+  lib,
+}: let
+  inherit (lib) licenses maintainers;
+in
+  stdenv.mkDerivation {
+    pname = "nss-altfiles";
+    version = "unstable-2020-09-25";
+
+    src = fetchFromGitHub {
+      owner = "flatcar";
+      repo = "nss-altfiles";
+      rev = "9078c543ba7d2bc5011737675b3dddb882673ce7";
+      sha256 = "sha256-mkZtuUsahHcwcmXvdH2thhDP7ctT5/wDpd0YUSSfd5w=";
+    };
+
+    configureFlags = [
+      "--with-types=hosts"
+      "--with-module-name='vula'"
+      "--datadir=/var/lib/vula-organize/"
+    ];
+
+    meta = {
+      description = "NSS module for relocating default file locations, tailored for Flatcar Container Linux";
+      homepage = "https://github.com/flatcar/nss-altfiles";
+      license = licenses.lgpl21Only;
+      maintainers = with maintainers; [mightyiam];
+    };
+  }

--- a/projects/Vula/service.nix
+++ b/projects/Vula/service.nix
@@ -7,71 +7,208 @@
 }: let
   inherit
     (lib)
+    assertMsg
+    getAttr
+    getExe
+    head
     types
     mkIf
     mkEnableOption
     mkOption
+    mkOrder
     mkPackageOption
+    recursiveUpdate
+    ;
+  inherit
+    (pkgs)
+    writeTextFile
+    coreutils
+    callPackage
     ;
 
   cfg = config.services.vula;
-  opt = options.services.vula;
+
+  nss-altfiles = callPackage ./nss-altfiles.nix {};
+
+  nssModuleName = "vula";
+
+  logLevelFlag =
+    getAttr
+    cfg.logLevel
+    {
+      INFO = "--info";
+      WARN = "--quiet";
+      DEBUG = "--verbose";
+    };
+
+  groupsAreUnique = assertMsg (cfg.operatorsGroup != cfg.systemGroup) "The options `config.services.vula.{systemGroup,operatorsGroup}` must have different values.";
+
+  commonServiceAttrs = {
+    serviceConfig = {
+      DevicePolicy = "closed";
+      Group = cfg.systemGroup;
+      LockPersonality = "yes";
+      MemoryDenyWriteExecute = "yes";
+      NoNewPrivileges = "yes";
+      PrivateDevices = "yes";
+      PrivateTmp = "yes";
+      ProtectControlGroups = "yes";
+      ProtectHome = "read-only";
+      ProtectKernelLogs = "yes";
+      ProtectKernelModules = "yes";
+      ProtectKernelTunables = "yes";
+      Restart = "always";
+      RestartSec = "5s";
+      RestrictNamespaces = "yes";
+      RestrictRealtime = "yes";
+      RestrictSUIDSGID = "yes";
+      Slice = "vula.slice";
+      StandardError = "journal";
+      StandardOutput = "journal";
+      Type = "dbus";
+    };
+    wantedBy = ["multi-user.target"];
+  };
 in {
-  options.services.vula = with types; {
-    enable = mkEnableOption "vula";
+  options.services.vula = {
+    enable = mkEnableOption "Enables Vula, \"automatic local network encryption\". The wireguard kernel module is required.";
 
     package = mkPackageOption pkgs "vula" {};
 
-    user = mkOption {
-      type = str;
-      description = "Username of the system user that should own files and services related to vula.";
+    userPrefix = mkOption {
+      type = types.str;
+      description = "Prefix for names of vula system users.";
       default = "vula";
     };
 
-    group = mkOption {
-      type = str;
-      description = "Group that contains the system user that executes vula.";
+    systemGroup = mkOption {
+      type = types.str;
+      description = "Group name for vula system users.";
       default = "vula";
+    };
+
+    operatorsGroup = mkOption {
+      type = types.str;
+      description = "Users in this group have full permissions to control vula.";
+      default = "vula-ops";
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      description = "Opens ports 5353 and 5354, and enables [option]`${options.networking.firewall.checkReversePath}`.";
+      default = false;
+    };
+
+    logLevel = mkOption {
+      type = types.enum ["INFO" "WARN" "DEBUG"];
+      description = "Vula daemons log level.";
+      default = "INFO";
+      example = "WARN";
     };
   };
 
   config = mkIf cfg.enable {
-    users.users."${cfg.user}" = {
+    system.nssModules = ["${nss-altfiles}"];
+    system.nssDatabases.hosts = mkOrder 0 [nssModuleName];
+
+    users.groups."${assert groupsAreUnique; cfg.systemGroup}" = {};
+    users.groups."${assert groupsAreUnique; cfg.operatorsGroup}" = {};
+
+    users.users."${cfg.userPrefix}-discover" = {
       isSystemUser = true;
-      group = cfg.group;
+      group = cfg.systemGroup;
     };
 
-    users.users."${cfg.user}-discover" = {
+    users.users."${cfg.userPrefix}-publish" = {
       isSystemUser = true;
-      group = cfg.group;
+      group = cfg.systemGroup;
     };
 
-    users.users."${cfg.user}-discover-alt" = {
+    users.users."${cfg.userPrefix}-organize" = {
       isSystemUser = true;
-      group = cfg.group;
+      group = cfg.systemGroup;
     };
-
-    users.users."${cfg.user}-publish" = {
-      isSystemUser = true;
-      group = cfg.group;
-    };
-
-    users.users."${cfg.user}-publish-alt" = {
-      isSystemUser = true;
-      group = cfg.group;
-    };
-
-    users.users."${cfg.user}-organize" = {
-      isSystemUser = true;
-      group = cfg.group;
-    };
-
-    users.groups."${cfg.group}" = {};
 
     environment.systemPackages = [cfg.package];
 
-    systemd.packages = [cfg.package];
+    services.dbus.packages =
+      [
+        (writeTextFile {
+          name = "vula-dbus.conf";
+          destination = "/share/dbus-1/system.d/local.vula.services.conf";
+          text = import ./dbus.conf.nix {inherit (cfg) userPrefix operatorsGroup;};
+        })
+      ]
+      ++ (map
+        (name:
+          writeTextFile {
+            name = "local.vula.${name}.service";
+            destination = "/share/dbus-1/system-services/local.vula.${name}.service";
+            text = ''
+              [D-BUS Service]
+              Name=local.vula.${name}
+              Exec=${coreutils}/bin/false
+              User=${cfg.userPrefix}-${name}
+              SystemdService=vula-${name}.service
+            '';
+          })
+        ["organize" "discover" "publish"]);
 
-    services.dbus.packages = [cfg.package];
+    networking.firewall.allowedUDPPorts = mkIf cfg.openFirewall [
+      5353 # mdns
+      5354 # port vula uses wireguard with
+    ];
+
+    networking.firewall.checkReversePath = mkIf cfg.openFirewall "loose";
+
+    systemd.services.vula-organize = recursiveUpdate commonServiceAttrs {
+      description = "Vula organize service daemon";
+      after = ["network.target" "vula-discover.target" "vula-publish.target"];
+      serviceConfig.AmbientCapabilities = "CAP_NET_ADMIN";
+      serviceConfig.BusName = "local.vula.organize";
+      serviceConfig.CapabilityBoundingSet = "CAP_NET_ADMIN";
+      serviceConfig.ExecStart = "${getExe cfg.package} ${logLevelFlag} organize";
+      serviceConfig.IPAddressDeny = "any";
+      serviceConfig.RestrictAddressFamilies = "AF_UNIX AF_NETLINK";
+      serviceConfig.StateDirectory = "vula-organize";
+      serviceConfig.TasksMax = "24";
+      serviceConfig.User = "${cfg.userPrefix}-organize";
+    };
+
+    systemd.services.vula-discover = recursiveUpdate commonServiceAttrs {
+      description = "Vula discover service daemon";
+      after = ["network.target"];
+      partOf = ["vula-discover.target"];
+      serviceConfig.BusName = "local.vula.discover";
+      serviceConfig.ExecStart = "${getExe cfg.package} ${logLevelFlag} discover";
+      serviceConfig.IPAddressAllow = "multicast";
+      serviceConfig.RestrictAddressFamilies = "AF_INET AF_INET6 AF_UNIX AF_NETLINK";
+      serviceConfig.TasksMax = "24";
+      serviceConfig.User = "${cfg.userPrefix}-discover";
+    };
+
+    systemd.services.vula-publish = recursiveUpdate commonServiceAttrs {
+      description = "Vula publish service daemon";
+      after = ["network.target"];
+      partOf = ["vula-publish.target"];
+      serviceConfig.BusName = "local.vula.publish";
+      serviceConfig.ExecStart = "${getExe cfg.package} ${logLevelFlag} publish";
+      serviceConfig.IPAddressAllow = "multicast";
+      serviceConfig.RestrictAddressFamilies = "AF_INET AF_INET6 AF_UNIX AF_NETLINK";
+      serviceConfig.TasksMax = "24";
+      serviceConfig.User = "${cfg.userPrefix}-publish";
+    };
+
+    systemd.slices.vula.description = "Slice for vula services";
+    systemd.slices.vula.before = ["slices.target"];
+    systemd.slices.vula.sliceConfig.MemoryMax = "512M";
+    systemd.slices.vula.sliceConfig.TasksMax = "72";
+
+    assertions = [
+      {
+        assertion = (head config.system.nssDatabases.hosts) == nssModuleName;
+        message = "Vula name security requires that the first value in `config.system.nssDatabases.hosts` is `\"vula\"`.";
+      }
+    ];
   };
 }

--- a/projects/Vula/test.nix
+++ b/projects/Vula/test.nix
@@ -1,21 +1,85 @@
 {
   sources,
-  pkgs,
+  lib,
   ...
-}: {
+}: let
+  inherit (lib) recursiveUpdate mkForce;
+in {
   name = "vula";
 
-  nodes = {
-    server = {config, ...}: {
-      imports = [
-        sources.modules.default
-        sources.modules."services.vula"
-      ];
-      services.vula.enable = true;
-    };
+  nodes.a = {
+    imports = [
+      sources.modules.default
+      sources.modules."services.vula"
+    ];
+
+    services.vula.enable = true;
+    services.vula.openFirewall = true;
+
+    services.vula.logLevel = "DEBUG";
+
+    # Make sure that if hosts can resolve each others' names,
+    # it is thanks to vula and the nss module it uses.
+    networking.extraHosts = mkForce "";
+
+    services.vula.userPrefix = "non-default-prefix";
+    services.vula.systemGroup = "non-default-system-group";
+    services.vula.operatorsGroup = "vula-managers";
+    users.users.user.isNormalUser = true;
+    users.users.admin.isNormalUser = true;
+    users.users.admin.extraGroups = ["vula-managers"];
   };
 
-  testScript = {nodes, ...}: ''
+  nodes.b.imports = [
+    sources.modules.default
+    sources.modules."services.vula"
+    ./example-simple.nix
+  ];
+
+  testScript = ''
     start_all()
+    a.wait_for_unit("vula-organize.service")
+    a.wait_for_unit("vula-publish.service")
+    a.wait_for_unit("vula-discover.service")
+    b.wait_for_unit("vula-organize.service")
+    b.wait_for_unit("vula-publish.service")
+    b.wait_for_unit("vula-discover.service")
+
+    def test_peer(node):
+        peer = 'a.local.' if node.name == 'b' else 'b.local.'
+        node.wait_until_succeeds(f"ping -I vula -c 1 {peer}", timeout=60)
+
+        peer_ip = node.succeed(f"getent hosts {peer}").split()[0]
+        route_result = node.succeed(f"ip route get {peer_ip}")
+        assert " dev vula " in route_result
+
+    test_peer(a)
+    test_peer(b)
+
+    a.succeed("pgrep --uid non-default-prefix-organize vula")
+    a.succeed("pgrep --uid non-default-prefix-discover vula")
+    a.succeed("pgrep --uid non-default-prefix-publish vula")
+
+    group_count = a.succeed("pgrep --count --group non-default-system-group vula").strip()
+    assert group_count == "3", "vula process group count should be 3"
+
+    # log level
+    a.succeed("pgrep --full -- 'vula.* --verbose organize'")
+    a.succeed("pgrep --full -- 'vula.* --verbose discover'")
+    a.succeed("pgrep --full -- 'vula.* --verbose publish'")
+
+    a.fail("su - user -c 'vula status'")
+    a.succeed("su - admin -c 'vula status'")
+
+    a.fail("su - user -c 'vula peer'")
+    a.succeed("su - admin -c 'vula peer'")
+
+    a.fail("su - user -c 'vula prefs set pin_new_peers true'")
+    a.succeed("su - admin -c 'vula prefs set pin_new_peers true'")
   '';
+
+  interactive.nodes.b = {
+    users.users.admin.isNormalUser = true;
+    users.users.admin.extraGroups = ["vula-admins"];
+  };
 }

